File: C#/src/BinaryDataPacker.cs
using System.Buffers.Binary;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.IO.Hashing;
using System.Linq;

namespace BinaryDataPacker;

public sealed class PacketConfigurationException : InvalidOperationException
{
    public PacketConfigurationException(string message) : base(message)
    {
    }
}

public sealed class PacketValidationException : InvalidOperationException
{
    public PacketValidationException(string message) : base(message)
    {
    }
}

public interface IHeaderBuilder
{
    int Size { get; }

    byte[] Build(ReadOnlySpan<byte> rawPayload);

    Action<byte[]> CreateValidator(ReadOnlySpan<byte> builtBytes);
}

public interface IPacketBuilder
{
    byte[] Apply(ReadOnlySpan<byte> data);

    byte[] Revert(ReadOnlySpan<byte> data);
}

public sealed class ConstantHeaderBuilder : IHeaderBuilder
{
    private readonly byte[] _value;

    public ConstantHeaderBuilder(ReadOnlySpan<byte> value)
    {
        if (value.IsEmpty)
        {
            throw new PacketConfigurationException("constant header value must not be empty");
        }

        _value = value.ToArray();
    }

    public int Size => _value.Length;

    public byte[] Build(ReadOnlySpan<byte> rawPayload) => _value;

    public Action<byte[]> CreateValidator(ReadOnlySpan<byte> builtBytes)
    {
        if (!builtBytes.SequenceEqual(_value))
        {
            throw new PacketValidationException("constant header/footer mismatch");
        }

        return _ => { };
    }
}

public sealed class PayloadLengthBuilder : IHeaderBuilder
{
    public int Size => 4;

    public byte[] Build(ReadOnlySpan<byte> rawPayload)
    {
        var buffer = new byte[Size];
        BinaryPrimitives.WriteInt32BigEndian(buffer, rawPayload.Length);
        return buffer;
    }

    public Action<byte[]> CreateValidator(ReadOnlySpan<byte> builtBytes)
    {
        if (builtBytes.Length != Size)
        {
            throw new PacketValidationException("invalid payload length field size");
        }

        var expectedLength = BinaryPrimitives.ReadInt32BigEndian(builtBytes);
        return payload =>
        {
            if (payload.Length != expectedLength)
            {
                throw new PacketValidationException(
                    $"payload length mismatch: expected {expectedLength}, got {payload.Length}");
            }
        };
    }
}

public sealed class Crc32Builder : IHeaderBuilder
{
    public int Size => 4;

    public byte[] Build(ReadOnlySpan<byte> rawPayload)
    {
        var checksum = Crc32.Hash(rawPayload);
        if (checksum.Length != Size)
        {
            throw new PacketConfigurationException("unexpected CRC-32 output size");
        }

        return checksum;
    }

    public Action<byte[]> CreateValidator(ReadOnlySpan<byte> builtBytes)
    {
        if (builtBytes.Length != Size)
        {
            throw new PacketValidationException("invalid CRC field size");
        }

        var expected = BinaryPrimitives.ReadUInt32BigEndian(builtBytes);
        return payload =>
        {
            var actual = BinaryPrimitives.ReadUInt32BigEndian(Crc32.Hash(payload));
            if (actual != expected)
            {
                throw new PacketValidationException(
                    $"CRC mismatch: expected 0x{expected:X8}, got 0x{actual:X8}");
            }
        };
    }
}

public sealed class XorPacketBuilder : IPacketBuilder
{
    private readonly byte[] _key;

    public XorPacketBuilder(ReadOnlySpan<byte> key)
    {
        if (key.IsEmpty)
        {
            throw new PacketConfigurationException("XOR key must not be empty");
        }

        _key = key.ToArray();
    }

    public byte[] Apply(ReadOnlySpan<byte> data) => Xor(data);

    public byte[] Revert(ReadOnlySpan<byte> data) => Xor(data);

    private byte[] Xor(ReadOnlySpan<byte> data)
    {
        var result = new byte[data.Length];
        for (var i = 0; i < data.Length; i++)
        {
            result[i] = (byte)(data[i] ^ _key[i % _key.Length]);
        }

        return result;
    }
}

public sealed class DeflatePacketBuilder : IPacketBuilder
{
    private readonly CompressionLevel _level;

    public DeflatePacketBuilder(CompressionLevel level = CompressionLevel.Optimal)
    {
        _level = level;
    }

    public byte[] Apply(ReadOnlySpan<byte> data)
    {
        using var output = new MemoryStream();
        using (var deflate = new DeflateStream(output, _level, leaveOpen: true))
        {
            deflate.Write(data);
        }

        return output.ToArray();
    }

    public byte[] Revert(ReadOnlySpan<byte> data)
    {
        using var input = new MemoryStream(data.ToArray());
        using var deflate = new DeflateStream(input, CompressionMode.Decompress);
        using var output = new MemoryStream();

        try
        {
            deflate.CopyTo(output);
        }
        catch (InvalidDataException ex)
        {
            throw new PacketValidationException($"failed to decompress payload: {ex.Message}");
        }

        return output.ToArray();
    }
}

public sealed class BinaryDataPacker
{
    private IHeaderBuilder? _sofBuilder;
    private readonly List<IHeaderBuilder> _eofBuilders = new();
    private readonly List<IPacketBuilder> _packetBuilders = new();

    public void AddSOFBuilder(IHeaderBuilder headerBuilder)
    {
        _sofBuilder = headerBuilder ?? throw new ArgumentNullException(nameof(headerBuilder));
    }

    public void AddEOFBuilder(IHeaderBuilder headerBuilder)
    {
        if (headerBuilder is null)
        {
            throw new ArgumentNullException(nameof(headerBuilder));
        }

        _eofBuilders.Add(headerBuilder);
    }

    public void AddPacketBuilder(IPacketBuilder packetBuilder)
    {
        if (packetBuilder is null)
        {
            throw new ArgumentNullException(nameof(packetBuilder));
        }

        _packetBuilders.Add(packetBuilder);
    }

    public byte[] GetPacket(byte[] rawData)
    {
        if (rawData is null)
        {
            throw new ArgumentNullException(nameof(rawData));
        }

        return BuildPacket(rawData);
    }

    public void GetPacket(byte[] raw_data, int size, ref byte[] result_buffer, ref int result_size)
    {
        if (raw_data is null)
        {
            throw new ArgumentNullException(nameof(raw_data));
        }

        if (size < 0 || size > raw_data.Length)
        {
            throw new ArgumentOutOfRangeException(nameof(size));
        }

        var payload = raw_data.AsSpan(0, size).ToArray();
        var packet = BuildPacket(payload);
        result_buffer = packet;
        result_size = packet.Length;
    }

    public byte[] UnpackData(byte[] packedData)
    {
        if (packedData is null)
        {
            throw new ArgumentNullException(nameof(packedData));
        }

        return ExtractPayload(packedData);
    }

    public void UnpackData(byte[] packed_data, int packet_size, ref byte[] result_buffer, ref int result_size)
    {
        if (packed_data is null)
        {
            throw new ArgumentNullException(nameof(packed_data));
        }

        if (packet_size < 0 || packet_size > packed_data.Length)
        {
            throw new ArgumentOutOfRangeException(nameof(packet_size));
        }

        var packet = packed_data.AsSpan(0, packet_size).ToArray();
        var payload = ExtractPayload(packet);
        result_buffer = payload;
        result_size = payload.Length;
    }

    private byte[] BuildPacket(ReadOnlySpan<byte> rawPayload)
    {
        var headerValidators = new List<Action<byte[]>>();
        var headerBytes = Array.Empty<byte>();

        if (_sofBuilder is not null)
        {
            headerBytes = _sofBuilder.Build(rawPayload);
            if (headerBytes.Length != _sofBuilder.Size)
            {
                throw new PacketConfigurationException("SOF builder returned unexpected number of bytes");
            }

            headerValidators.Add(_sofBuilder.CreateValidator(headerBytes));
        }

        var footerBytes = new List<byte>();
        var footerValidators = new List<Action<byte[]>>();
        foreach (var builder in _eofBuilders)
        {
            var bytes = builder.Build(rawPayload);
            if (bytes.Length != builder.Size)
            {
                throw new PacketConfigurationException("EOF builder returned unexpected number of bytes");
            }

            footerBytes.AddRange(bytes);
            footerValidators.Add(builder.CreateValidator(bytes));
        }

        var transformed = rawPayload.ToArray();
        foreach (var builder in _packetBuilders)
        {
            transformed = builder.Apply(transformed);
        }

        var packet = new byte[headerBytes.Length + 4 + transformed.Length + footerBytes.Count];
        var cursor = 0;

        Array.Copy(headerBytes, 0, packet, cursor, headerBytes.Length);
        cursor += headerBytes.Length;

        BinaryPrimitives.WriteInt32BigEndian(packet.AsSpan(cursor, 4), transformed.Length);
        cursor += 4;

        Array.Copy(transformed, 0, packet, cursor, transformed.Length);
        cursor += transformed.Length;

        if (footerBytes.Count > 0)
        {
            footerBytes.CopyTo(packet, cursor);
        }

        foreach (var validator in headerValidators)
        {
            validator(rawPayload.ToArray());
        }

        foreach (var validator in footerValidators)
        {
            validator(rawPayload.ToArray());
        }

        return packet;
    }

    private byte[] ExtractPayload(ReadOnlySpan<byte> packet)
    {
        var validators = new List<Action<byte[]>>();
        var cursor = 0;

        if (_sofBuilder is not null)
        {
            if (packet.Length < _sofBuilder.Size)
            {
                throw new PacketValidationException("packet is too small to contain the header");
            }

            var header = packet.Slice(cursor, _sofBuilder.Size).ToArray();
            validators.Add(_sofBuilder.CreateValidator(header));
            cursor += _sofBuilder.Size;
        }

        var footerSegments = new List<Action<byte[]>>();
        if (_eofBuilders.Count > 0)
        {
            var remaining = packet.Length - cursor;
            for (var index = _eofBuilders.Count - 1; index >= 0; index--)
            {
                var builder = _eofBuilders[index];
                if (remaining < builder.Size)
                {
                    throw new PacketValidationException("packet is too small to contain the footer");
                }

                var segmentStart = packet.Length - builder.Size;
                var footerSegment = packet.Slice(segmentStart, builder.Size).ToArray();
                footerSegments.Add(builder.CreateValidator(footerSegment));
                packet = packet.Slice(0, segmentStart);
                remaining = packet.Length - cursor;
            }

            footerSegments.Reverse();
            validators.AddRange(footerSegments);
        }

        if (packet.Length - cursor < 4)
        {
            throw new PacketValidationException("packet does not contain body length information");
        }

        var bodyLength = BinaryPrimitives.ReadInt32BigEndian(packet.Slice(cursor, 4));
        cursor += 4;

        var body = packet.Slice(cursor).ToArray();
        if (body.Length != bodyLength)
        {
            throw new PacketValidationException(
                $"body length mismatch: declared {bodyLength}, actual {body.Length}");
        }

        var transformed = body;
        for (var index = _packetBuilders.Count - 1; index >= 0; index--)
        {
            transformed = _packetBuilders[index].Revert(transformed);
        }

        foreach (var validator in validators)
        {
            validator(transformed);
        }

        return transformed;
    }
}

File: Python/src/BinaryDataPacker.py
"""Binary data packetization utilities.

This module implements a simple but extensible packetization system that
allows callers to decorate raw payloads with headers, footers and a chain
of transformations (encryption, compression, ...).  The implementation is
purposely lightweight and only depends on the Python standard library so it
can be used in resource constrained environments as well as in unit tests.

The design is intentionally declarative.  A :class:`BinaryDataPacker`
instance is configured with builders that produce fixed size header/footer
segments based on the *raw* payload.  The packer also maintains a
collection of :class:`PacketBuilder` instances that are applied sequentially
when the payload is packed and rolled back in reverse order during
unpacking.  Each builder contributes an optional validator that is executed
once the raw payload has been reconstructed.

Even though the original project description mentions a huge amount of
combinations, the Python implementation focuses on a pragmatic and tested
subset that covers the most common scenarios: constant markers, payload
length information, CRC32 checksums, XOR based encryption and DEFLATE
compression.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, List, Optional
import binascii
import zlib


class PacketConfigurationError(ValueError):
    """Raised when the packer is misconfigured."""


class PacketValidationError(ValueError):
    """Raised when validation fails while unpacking a packet."""


Validator = Callable[[bytes], None]


class FixedSizeBuilder:
    """Base class for header/footer builders with a fixed output size."""

    size: int

    def __init__(self, size: int) -> None:
        if size <= 0:
            raise PacketConfigurationError("builder size must be positive")
        self.size = size

    def build(self, raw_payload: bytes) -> bytes:
        raise NotImplementedError

    def create_validator(self, built_bytes: bytes) -> Validator:
        raise NotImplementedError

    def __repr__(self) -> str:  # pragma: no cover - helper for debugging
        return f"{self.__class__.__name__}(size={self.size})"


class ConstantBuilder(FixedSizeBuilder):
    """Builder that always emits the same byte sequence."""

    def __init__(self, value: bytes) -> None:
        if not isinstance(value, (bytes, bytearray)):
            raise PacketConfigurationError("constant value must be bytes")
        self._value = bytes(value)
        super().__init__(len(self._value))

    def build(self, raw_payload: bytes) -> bytes:  # pragma: no cover - trivial
        return self._value

    def create_validator(self, built_bytes: bytes) -> Validator:
        expected = self._value

        if built_bytes != expected:
            raise PacketValidationError("constant header/footer mismatch")

        return lambda _: None


class PayloadLengthBuilder(FixedSizeBuilder):
    """Builder that encodes the raw payload length as an unsigned integer."""

    def __init__(self) -> None:
        super().__init__(size=4)

    def build(self, raw_payload: bytes) -> bytes:
        return len(raw_payload).to_bytes(self.size, byteorder="big")

    def create_validator(self, built_bytes: bytes) -> Validator:
        if len(built_bytes) != self.size:
            raise PacketValidationError("invalid payload length field size")
        expected_length = int.from_bytes(built_bytes, byteorder="big")

        def _validate(raw_payload: bytes) -> None:
            if len(raw_payload) != expected_length:
                raise PacketValidationError(
                    "payload length mismatch: expected "
                    f"{expected_length}, got {len(raw_payload)}"
                )

        return _validate


class CRC32Builder(FixedSizeBuilder):
    """Builder that stores a CRC32 checksum of the raw payload."""

    def __init__(self) -> None:
        super().__init__(size=4)

    def build(self, raw_payload: bytes) -> bytes:
        checksum = binascii.crc32(raw_payload) & 0xFFFFFFFF
        return checksum.to_bytes(self.size, byteorder="big")

    def create_validator(self, built_bytes: bytes) -> Validator:
        if len(built_bytes) != self.size:
            raise PacketValidationError("invalid CRC field size")
        expected = int.from_bytes(built_bytes, byteorder="big")

        def _validate(raw_payload: bytes) -> None:
            actual = binascii.crc32(raw_payload) & 0xFFFFFFFF
            if actual != expected:
                raise PacketValidationError(
                    "CRC mismatch: expected "
                    f"0x{expected:08X}, got 0x{actual:08X}"
                )

        return _validate


class PacketBuilder:
    """Base class for transformations applied to payloads."""

    def apply(self, data: bytes) -> bytes:
        raise NotImplementedError

    def revert(self, data: bytes) -> bytes:
        raise NotImplementedError


@dataclass
class XorCipher(PacketBuilder):
    """Symmetric XOR based cipher using a repeating key."""

    key: bytes

    def __post_init__(self) -> None:
        if not self.key:
            raise PacketConfigurationError("XOR key must not be empty")
        if not isinstance(self.key, (bytes, bytearray)):
            raise PacketConfigurationError("XOR key must be bytes")
        self.key = bytes(self.key)

    def _xor(self, data: bytes) -> bytes:
        key = self.key
        key_len = len(key)
        return bytes(b ^ key[i % key_len] for i, b in enumerate(data))

    def apply(self, data: bytes) -> bytes:
        return self._xor(data)

    def revert(self, data: bytes) -> bytes:
        return self._xor(data)


class DeflateCompressor(PacketBuilder):
    """Applies DEFLATE compression and decompression using zlib."""

    def __init__(self, level: int = 6) -> None:
        if not (0 <= level <= 9):
            raise PacketConfigurationError("compression level must be between 0 and 9")
        self.level = level

    def apply(self, data: bytes) -> bytes:
        return zlib.compress(data, self.level)

    def revert(self, data: bytes) -> bytes:
        try:
            return zlib.decompress(data)
        except zlib.error as exc:  # pragma: no cover - exercised via tests
            raise PacketValidationError(f"failed to decompress payload: {exc}") from exc


class BinaryDataPacker:
    """Utility that packs and unpacks binary payloads."""

    def __init__(self) -> None:
        self._sof_builder: Optional[FixedSizeBuilder] = None
        self._eof_builders: List[FixedSizeBuilder] = []
        self._packet_builders: List[PacketBuilder] = []

    # ------------------------------------------------------------------
    # Configuration helpers
    def set_sof_builder(self, builder: FixedSizeBuilder) -> None:
        if not isinstance(builder, FixedSizeBuilder):
            raise PacketConfigurationError("SOF builder must derive from FixedSizeBuilder")
        self._sof_builder = builder

    def add_eof_builder(self, builder: FixedSizeBuilder) -> None:
        if not isinstance(builder, FixedSizeBuilder):
            raise PacketConfigurationError("EOF builder must derive from FixedSizeBuilder")
        self._eof_builders.append(builder)

    def clear_eof_builders(self) -> None:
        self._eof_builders.clear()

    def add_packet_builder(self, builder: PacketBuilder) -> None:
        if not isinstance(builder, PacketBuilder):
            raise PacketConfigurationError("packet builder must derive from PacketBuilder")
        self._packet_builders.append(builder)

    def clear_packet_builders(self) -> None:
        self._packet_builders.clear()

    # ------------------------------------------------------------------
    @staticmethod
    def _ensure_bytes(data: bytes | bytearray | memoryview | str) -> bytes:
        if isinstance(data, bytes):
            return data
        if isinstance(data, bytearray):
            return bytes(data)
        if isinstance(data, memoryview):
            return data.tobytes()
        if isinstance(data, str):
            return data.encode("utf-8")
        raise TypeError("payload must be bytes, bytearray, memoryview or str")

    def _build_header(self, raw_payload: bytes) -> tuple[bytes, List[Validator]]:
        if self._sof_builder is None:
            return b"", []
        header_bytes = self._sof_builder.build(raw_payload)
        if len(header_bytes) != self._sof_builder.size:
            raise PacketConfigurationError(
                "SOF builder returned an unexpected number of bytes"
            )
        validator = self._sof_builder.create_validator(header_bytes)
        return header_bytes, [validator]

    def _build_footer(self, raw_payload: bytes) -> tuple[bytes, List[Validator]]:
        footer_bytes = bytearray()
        validators: List[Validator] = []
        for builder in self._eof_builders:
            segment = builder.build(raw_payload)
            if len(segment) != builder.size:
                raise PacketConfigurationError(
                    f"builder {builder} returned an unexpected number of bytes"
                )
            footer_bytes.extend(segment)
            validators.append(builder.create_validator(segment))
        return bytes(footer_bytes), validators

    def _extract_header(
        self, packet: memoryview
    ) -> tuple[memoryview, List[Validator]]:
        if self._sof_builder is None:
            return packet, []
        size = self._sof_builder.size
        if len(packet) < size:
            raise PacketValidationError("packet is too small to contain the header")
        header = packet[:size]
        validator = self._sof_builder.create_validator(header.tobytes())
        return packet[size:], [validator]

    def _extract_footer(
        self, packet: memoryview
    ) -> tuple[memoryview, List[Validator]]:
        if not self._eof_builders:
            return packet, []
        validators: List[Validator] = []
        end = len(packet)
        for builder in reversed(self._eof_builders):
            size = builder.size
            if end < size:
                raise PacketValidationError("packet is too small to contain the footer")
            segment = packet[end - size : end]
            validators.append(builder.create_validator(segment.tobytes()))
            end -= size
        return packet[:end], list(reversed(validators))

    # ------------------------------------------------------------------
    def get_packet(self, payload: bytes | bytearray | memoryview | str) -> bytes:
        """Pack *payload* into a binary packet."""

        raw_payload = self._ensure_bytes(payload)
        header, header_validators = self._build_header(raw_payload)
        footer, footer_validators = self._build_footer(raw_payload)

        transformed = raw_payload
        for builder in self._packet_builders:
            transformed = builder.apply(transformed)

        body_length = len(transformed).to_bytes(4, byteorder="big")
        packet = bytearray()
        packet.extend(header)
        packet.extend(body_length)
        packet.extend(transformed)
        packet.extend(footer)

        # Validators are only used during unpacking but building them here
        # ensures that potential configuration errors are caught early.
        for validator in header_validators + footer_validators:
            validator(raw_payload)

        return bytes(packet)

    def unpack_data(self, packet: bytes | bytearray | memoryview) -> bytes:
        """Unpack *packet* and return the original payload."""

        if not isinstance(packet, (bytes, bytearray, memoryview)):
            raise TypeError("packet must be bytes-like")

        view = memoryview(packet if isinstance(packet, (bytes, bytearray)) else packet.tobytes())

        view, header_validators = self._extract_header(view)
        view, footer_validators = self._extract_footer(view)

        if len(view) < 4:
            raise PacketValidationError("packet does not contain body length information")
        body_length = int.from_bytes(view[:4], byteorder="big")
        body = view[4:]
        if len(body) != body_length:
            raise PacketValidationError(
                "body length mismatch: declared "
                f"{body_length}, actual {len(body)}"
            )

        transformed = body.tobytes()
        for builder in reversed(self._packet_builders):
            transformed = builder.revert(transformed)

        for validator in header_validators + footer_validators:
            validator(transformed)

        return transformed


__all__ = [
    "BinaryDataPacker",
    "CRC32Builder",
    "ConstantBuilder",
    "DeflateCompressor",
    "PacketConfigurationError",
    "PacketValidationError",
    "PayloadLengthBuilder",
    "XorCipher",
]
